src/io/save_gals_hdf5.c

```diff
@@ -0,0 +1,1149 @@
+    CHECK_STATUS_AND_RETURN_ON_FAIL(status, (int32_t) status, \
+                                    "Failed to write a dataset for field #field_name.\n" \
+                                    "The dimensions of the dataset was %d\nThe file ID was %d\n." \
+                                    "The dataset ID was %d.", (int32_t) dims[0], (int32_t) file_id, \
+                                    (int32_t) macro_dataset_id); \
+    status = H5Dclose(macro_dataset_id); \
+    CHECK_STATUS_AND_RETURN_ON_FAIL(status, (int32_t) status, \
+                                    "Failed to close the dataset for field #field_name.\n" \
+                                    "The dimensions of the dataset was %d\nThe file ID was %d\n." \
+                                    "The dataset ID was %d.", (int32_t) dims[0], (int32_t) file_id, \
+                                    (int32_t) macro_dataset_id); \
+    status = H5Sclose(macro_dataspace_id); \
+    CHECK_STATUS_AND_RETURN_ON_FAIL(status, (int32_t) status, \
+                                    "Failed to close the dataspace for field #field_name.\n" \
+                                    "The dimensions of the dataset was %d\nThe file ID was %d\n." \
+                                    "The dataspace ID was %d.", (int32_t) dims[0], (int32_t) file_id, \
+                                    (int32_t) macro_dataspace_id);\
+}
+
+// Unlike the binary output where we generate an array of output struct instances, the HDF5 workflow has
+// a single output struct (for each snapshot) where the **properties** of the struct are arrays.
+// This macro callocs (i.e., allocates and zeros) space for these inner arrays.
+#define CALLOC_GALAXY_OUTPUT_INNER_ARRAY(snap_idx, field_name) {     \
+    save_info->buffer_output_gals[snap_idx].field_name = calloc(save_info->buffer_size, sizeof(*(save_info->buffer_output_gals[snap_idx].field_name))); \
+    if(save_info->buffer_output_gals[snap_idx].field_name == NULL) { \
+        fprintf(stderr, "Could not allocate %d elements for the #field_name GALAXY_OUTPUT field for output snapshot #snap_idx\n", save_info->buffer_size);\
+        return MALLOC_FAILURE;                                       \
+    }                                                                \
+}
+
+#define FREE_GALAXY_OUTPUT_INNER_ARRAY(snap_idx, field_name) {     \
+    free(save_info->buffer_output_gals[snap_idx].field_name);      \
+}
+
+// Externally Visible Functions //
+
+// Creates the HDF5 file, groups and the datasets.  The heirachy for the HDF5 file is
+// File->Group->Datasets.  For example, File->"Snap_43"->"StellarMass"->**Data**.
+// The handles for all of these are stored in `save_info` so we can write later.
+int32_t initialize_hdf5_galaxy_files(const int filenr, struct save_info *save_info, const struct params *run_params)
+{
+    hid_t prop, dataset_id;
+    hid_t file_id, group_id, dataspace_id;
+    char buffer[3*MAX_STRING_LEN];
```

Comment:
I would have that this would require `[3*MAX_STRING_LEN + 5/6]` or trigger a compiler warning otherwise

Reply:
Ahh because `filenr` is an integer and contains many less digits than `MAX_STRING_LEN`